### PR TITLE
Boundary cons

### PR DESCRIPTION
**Pull request by alsgregory**
Alterations:

- Added `boundary_conditions` class which user can specify which bcs for each marked boundary are implemented through `flux.py`.
- Defaults each boundary to `solid wall` conditions unless specified.
- Added test of well balanced conditions given bcs of manual setup of `solid wall` conditions with `river` bc specified.
- Added test of all defaulted boundaries

To do before merge:

- [x] Create an `outflow` option with info RE Onno's email, changing `river` to `inflow`

### CODE REVIEW

File: flooddrake/timestepper.py [hunk 1]
```diff
@@ -38,6 +40,33 @@ def __init__(self, V, bed, source, MaxTimestep=0.025, func=lambda x: 1):
...
+        self.boundary_conditions = boundary_conditions
+        if self.boundary_conditions is None:
+            if self.mesh.geometric_dimension() == 1:
+                self.boundary_conditions = [BoundaryConditions(1),
+                                            BoundaryConditions(2)]
+            if self.mesh.geometric_dimension() == 2:
+                self.boundary_conditions = [BoundaryConditions(1),
+                                            BoundaryConditions(2),
+                                            BoundaryConditions(3),
+                                            BoundaryConditions(4)]
```
Comment (colinjcotter) [inline flooddrake/timestepper.py]: Probably a better way to organise this is to ask the user to provide a list of tags and boundaryconditions they would like to apply, and then loop over them, adding into the flux one by one in the loop.

Comment (alsgregory) [inline flooddrake/timestepper.py]: Yes indeed there is. Boundary tags are shown in `mesh.topology.exterior_facets.unique_markers` so we can loop over these

Comment (alsgregory) [inline flooddrake/timestepper.py]: These can be taken from the mesh, so no need to ask the user. They specify the tags they want to apply BCs to, and we default the rest of the tags to `solid wall` conditions

**Comment by alsgregory** [comment]
@colinjcotter 

see commit: https://github.com/firedrakeproject/flooddrake/pull/15/commits/1f9f44afd25ca497ce12ff2a969f6e36b731abbc 

one can now use any mesh markers / tags - not just a square mesh

**Comment by alsgregory** [comment]
@colinjcotter How's this looking w.r.t. merging?

**Pull request merged**